### ISSUE TITLE
Unable to Enable apps in OwnCloud v8.1 on rPi 2

### ISSUE THREAD

**Issue by cmeans** [owncloud/core#18088]
OwnCloud v8.1.0 (Stable)

I get the following error:

"save_to must be a GuzzleHttp\Stream\StreamInterface or resource"

Whenever I attempt to enable an app via the Admin/Apps UI.

**Comment by karlitschek** [comment]
Can you post a full bug report using the issue template please? https://github.com/owncloud/core/blob/master/CONTRIBUTING.md

**Comment by cmeans** [comment]
### Steps to reproduce
1. Login to OwnCloud Web UI as Admin
2. Navigate to Apps page
3. Select Productivity group
4. Click Enable button for "Calendar 0.7.2"

### Expected behaviour
I would expect the App to be Enabled...I'm evaluating this software, so am a total noob.

### Actual behaviour
Error message displayed below button:

"save_to must be a GuzzleHttp\Stream\StreamInterface or resource"

### Server configuration

**Operating system**:
Linux 3.18.11-v7+ #781 SMP PREEMPT Tue Apr 21 18:07:59 BST 2015 armv7l 

**Web server:**
nginx/1.2.1

**Database:**
MySQL 5.5.44-0+deb7u1 (Debian)

**PHP version:**
5.4.41-0+deb7u1

**ownCloud version:** (see ownCloud admin page)
8.1.0 (stable) 

**Updated from an older ownCloud or fresh install:**
Fresh install.

**List of activated apps:**
Whatever is activated by Default

**Comment by Pingumann** [comment]
Same issue with ownCloud 8.2. on CentOS 6 Minimal System (64 Bit) VPS. 

uname: 2.6.32-042stab108.2

httpd: httpd -version
Server version: Apache/2.2.15 (Unix)
Server built:   Oct 16 2014 14:48:21

mySQL: 
mysqld  Ver 10.0.20-MariaDB for Linux on x86_64 (MariaDB Server)
Copyright (c) 2000, 2015, Oracle, MariaDB Corporation Ab and others.

PHP:
php --version
PHP 5.4.42 (cli) (built: Jun 10 2015 14:19:12)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2014 Zend Technologies

**Comment by cmeans** [comment]
Almost a week has gone by, and this bug report hasn't even been classified yet.  It's now been confirmed in at least two different versions and environments.

Do you need more information or is this just way down in priorities?  It's difficult to evaluate software that doesn't work right.

**Comment by LukasReschke** [comment]
Works fine here on a CentOS 7, @cmeans what distribution are you using?

**Comment by LukasReschke** [comment]
Also do you folks happen to have SELinux enabled? If so try disabling it. There may be a good chance that it is blocking remote requests and the code is not handling this gracefully yet.

**Comment by Pingumann** [comment]
SELinux is disabled on CentOS 6.7.

lsr@servr ~]$ sestatus
SELinux status:                 disabled

**Comment by LukasReschke** [comment]
Please post your config in this case as well. It is missing from https://github.com/owncloud/core/issues/18088#issuecomment-128222204 (see also https://raw.githubusercontent.com/owncloud/core/master/issue_template.md)

**Comment by Pingumann** [comment]
OK, will do.

:-)

**Comment by LukasReschke** [comment]
Seems to be thrown at https://github.com/owncloud/3rdparty/blob/5cd078af93210ecbc21714369f60df3bb625f52c/guzzlehttp/ringphp/src/Client/CurlFactory.php#L395-L419, which likely means that `is_resource` returns false. 

However it should not reach this code path since the `is_string` check does a `is_dir`, so it seems like there is a problem with the temporary files generation which is done at https://github.com/owncloud/core/blob/6e4a79f8529a69b846f73257cc581d2b80f7762d/lib/private/installer.php#L266-L271.

So `$path=OC_Helper::tmpFile('.' . $pathInfo['extension']);` seems to go wrong here. Interesting.

Can you folks in `lib/private/installer.php` below this line please add the following line of code:

```
\OC::$server->getLogger()->critical(sprintf('Debug for 18088: %s', print_r($path, true)), ['app' => 'core']);
```

Then please try to enable an app and report back what the output is there.

**Comment by Pingumann** [comment]
I added the line - no difference. 

Should PHP, Apache2 or MySQL be restarted to read the file?

**Comment by LukasReschke** [comment]
Did you try to install another app? is certainly nothing in the log?

**Comment by LukasReschke** [comment]
Restart might help if you use opcaching.

**Comment by Pingumann** [comment]
I tried to install various Apps; see screen-shot.

I read that opcaching is new in php 5.5 and onwards. My server has php 5.4. 

I rebooted the Server - no difference. Interestingly, if I deactivate an active App, it deactivates. Then I can also activate it again without problems.

![Uploading Screen Shot 2015-08-11 at 13.22.26.png…]()

**Comment by LukasReschke** [comment]
And what does the ownCloud log contains for entries?

**Comment by cmeans** [comment]
I reviewed the log (sorry I didn't think to do this earlier).

It indicated a rights issue with the "tmp" folder.  I checked and "root"
was the owner not "www-data".  Once I changed that, I am able to Enable
apps without error.

-Chris

**Comment by LukasReschke** [comment]
Ok. We need to add to ownCloud a check if it can write to the temporary folder.

**Comment by ymybe** [comment]
Here is my log. I also tried to apply your fix for lib/private/installer.php 
no success.

is there anything i could to to fix this before 8.2 ?
-W

### Steps to reproduce
1. OC 8.1.0 to 8.1.1 via sudo -u www-data php occ upgrade
2. Activate Apps via Menu
(had problems with app activation before but calendar and contacts are the only i really need..)

### Expected behaviour
The Apps are back working as before the upgrade

### Actual behaviour
App is not activated

save_to must be a GuzzleHttp\Stream\StreamInterface or resource

### Server configuration
**Operating system**:Linux/Debian 4.1.5

**Web server: Apache 2.4.16**

**Database: Mysql 5.5.43**

**PHP version: 5.6.12**

**ownCloud version: 8.1.1** (see ownCloud admin page)

**Updated from an older ownCloud or fresh install: upgrade from 8.1.0**

**List of activated apps:Activity, Deleted Files, Encryption, File Locking, First run wizzard, Mail Template Editor, PDF Viewer, Pictures, Share Files, Text editor, Updater, Videoviewer, Versions**


**The content of config/config.php: <?php
$CONFIG = array (
  'instanceid' => 'no',
  'passwordsalt' => 'no',
  'trusted_domains' => 
  array (
    0 => 'no',
  ),
  'datadirectory' => '/nas/cloud',
  'dbtype' => 'mysql',
  'version' => '8.1.1.3',
  'dbname' => 'no',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'no',
  'dbpassword' => 'no',
  'installed' => true,
  'mail_smtpmode' => 'smtp',
  'mail_smtpname' => 'no',
  'forcessl' => true,
  'mail_from_address' => 'c',
  'mail_domain' => 'no',
  'mail_smtppassword' => 'no',
  'mail_smtphost' => 'no',
  'loglevel' => 4,
  'mail_smtpauth' => true,
  'theme' => '',
  'maintenance' => false,
  'mail_smtpauthtype' => 'PLAIN',
  'mail_smtpsecure' => 'tls',
  'secret' => 'no',
  'forceSSLforSubdomains' => true,
  'singleuser' => false,
  'appstore.experimental.enabled' => false,
);
**

**Are you using external storage, if yes which one:** none

**Are you using encryption:** yes

**Are you using an external user-backend, if yes which one:**  none

### Logs


#### ownCloud log (data/owncloud.log)

Exception: {"Message":"HTTP\/1.1 503 Exception: App not installed: calendar","Code":0,"Trace":"#0 [internal function]: {closure}(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#1 \/var\/www\/cloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\n#2 \/var\/www\/cloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(456): Sabre\\Event\\EventEmitter->emit('beforeMethod', Array)\n#3 \/var\/www\/cloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#4 \/var\/www\/cloud\/remote.php(64): Sabre\\DAV\\Server->exec()\n#5 \/var\/www\/cloud\/remote.php(135): handleException(Object(Exception))\n#6 {main}","File":"\/var\/www\/cloud\/remote.php","Line":62}

**Comment by Pingumann** [comment]
I removed the edit to the installer.php, as it made no difference. I also checked the /var/tmp/ and /tmp/ directories - all standard rights. Nothing changed since original OS installation. CentOS has apache:apache to run webserver. Access rights are correct as well.

Then, in the owncloud.log I noticed the following error message:

{"reqId":"XSknjRUwhxyCu+E4PD21","remoteAddr":"xxx.xxx.xxx.xxx","app":"core","message":"save_to must be a GuzzleHttp\\Stream\\StreamInterface or resource","level":3,"time":"2015-08-11T20:01:05+00:00"}

Then a strange directory in /var/ it tries to create called "big_temp_file". 

{"reqId":"gETgWdod\/nrE\/7ueNMC2","remoteAddr":"xxx.xxx.xxx.xxx","app":"no app in context","message":"Can not create a temporary file in directory \/var\/big_temp_file\/. Check it exists and has correct permissions","level":2,"time":"2015-08-11T20:01:07+00:00"}

Why is it trying to create a tmp file under /var ? Should apache:apache rights be given the whole /var directory? Should it not be in /var/tmp/ or /tmp/ ? I'd think it should be /tmp/big_temp_file/ or /var/tmp/big_temp_file/.

**Comment by LukasReschke** [comment]
>Here is my log. I also tried to apply your fix for lib/private/installer.php 

That's because it is not a fix but only a debugging measure to get the actual error.

```
{"reqId":"gETgWdod\/nrE\/7ueNMC2","remoteAddr":"xxx.xxx.xxx.xxx","app":"no app in context","message":"Can not create a temporary file in directory \/var\/big_temp_file\/. Check it exists and has correct permissions","level":2,"time":"2015-08-11T20:01:07+00:00"}
```

That is interesting. It seems that `get_temp_dir()` returns some kind of wrong value then. Can you please upload the following file and let me know what the result is? 

```php
<?php
function get_temp_dir() {
	if ($temp = ini_get('upload_tmp_dir')) return '1'.$temp;
	if ($temp = getenv('TMP')) return '2'.$temp;
	if ($temp = getenv('TEMP')) return '3'.$temp;
	if ($temp = getenv('TMPDIR')) return '4'.$temp;
	$temp = tempnam(__FILE__, '');
	if (file_exists($temp)) {
		unlink($temp);
		return '5'.dirname($temp);
	}
	if ($temp = sys_get_temp_dir()) return '6'.$temp;

	return null;
}

var_dump(get_temp_dir());
```

**Comment by ymybe** [comment]
it says:

string(9) "1/tmp/php"

(i also tried to add:

  'memcache.local' => '\\OC\\Memcache\\Memcached',
  'cache_path' => '/tmp',

in my config.php while testing your script, while also reloading index.php no change only /tmp/php)

**Comment by karlitschek** [comment]
Yes. We should check for the right permissions. Who wants to volunteer? :-)

**Comment by ymybe** [comment]
Changing the php upload dir (php.ini) from /tmp/php to /tmp lets me enable the app but asks me to upgrade again (doing that via sudo -u www-data occ upgrade one app at a time...)

**Comment by Pingumann** [comment]
@karlitschek : I can check permissions. 

@ymybe : You changed /etc/php.ini? My php.ini has no /tmp/php, only a /tmp.

**Comment by Pingumann** [comment]
Will a new install with the current ownCloud Version 8.1.1 solve the problems?

I installed originally with the tarball from ownCloud website, then saw that it can also be installed with "yum install owncloud".

Doing that afterwards produces a profuse amount of dependency errors, wanting a previous php version than what is available on the system.

**Comment by Pingumann** [comment]
I did an upgrade with the ownCloud Admin option to upgrade. It went without a hitch; impressive!

Now I have ownCloud 8.1.1 Stable.

App activation has not changed. Issue is still present.

@karlitschek : So what next? I can do some permission checking. How do I proceed and what info do you require?

**Comment by karlitschek** [comment]
@Pingumann What I wanted to say is that we should add a check in ownCloud itself. Do you want to send a pull request? :-)

**Comment by ymybe** [comment]
sudo -u APACHEUSER touch /tmp/test
 (www-data with debian, dunno)
if there is a file in /tmp called test your webserver should work with that)

also 
user@server:/# ls -ahl /var/www/owncloud
total 148K
drwx------ 1 www-data www-data  450 Aug 12 08:53 .
(...)

(or where ever it is stored) it should state the rights (rw- --- --- or so) and the assigned user and group (www-data or your apache user) usually everything should work if its write & readable by your webserver

 from my experience i can recommend to install the tarball (and then dont forgett to change filesystem permissons: chmod, chown..) because owncloud has some depencies of course but _downgrading_ php isnt one of them..


[sorry if would find a messaging button on github i might would have used that for the support here...]

**Comment by Pingumann** [comment]
@karlitschek : I am too much of a novice for github yet for 'pull requesting’, & I am only a Bash scripter - no programmer. I will follow what @ymybe suggests.

@ymybe :OK, here are my findings: 

1: setting /tmp/test:
--------------------------
[user1@oc1 ~]$ sudo -u apache touch /tmp/test
[sudo] password for usr1:
[user1@oc1 ~]$ sudo ls -al /tmp/test
-rw-r--r-- 1 apache apache 0 Aug 15 08:57 /tmp/test

2. Permissions of owncloud directory:
---------------------------------------------------
I used the tarball.
I set the chown -R apache:apache /var/www/html/owncloud
Seems that I kept the original permissions, not using chmod.

Installed and currently running owncloud:

[user1@oc1 ~]$ ls -ahl /var/www/html/owncloud/
total 200K
drwxr-xr-x 13 apache apache 4.0K Aug 14 05:39 .
drwxr-xr-x 10 apache apache 4.0K Jul 24 06:53 ..
-rw-r--r--  1 apache apache 2.1K Aug 14 05:39 .htaccess
-rw-r--r--  1 apache apache  19K Aug 14 05:38 .mailmap
-rw-r--r--  1 apache apache   12 Aug 14 05:38 .tag
-rw-r--r--  1 apache apache  164 Aug 14 05:38 .user.ini
drwxr-xr-x 26 apache apache 4.0K Aug 14 05:38 3rdparty
-rw-r--r--  1 apache apache  477 Aug 14 05:38 AUTHORS
-rw-r--r--  1 apache apache  34K Aug 14 05:38 COPYING-AGPL
drwxr-xr-x 22 apache apache 4.0K Aug 14 05:38 apps
drwxr-xr-x  2 apache apache 4.0K Jul 28 09:51 config
-rw-r--r--  1 apache apache 2.5K Aug 14 05:38 console.php
drwxr-xr-x 19 apache apache 4.0K Aug 14 05:38 core
-rw-r--r--  1 apache apache 5.1K Aug 14 05:38 cron.php
drwxrwx---  6 apache apache 4.0K Aug 14 05:48 data
-rw-r--r--  1 apache apache  22K Aug 14 05:38 db_structure.xml
-rw-r--r--  1 apache apache  179 Aug 14 05:38 index.html
-rw-r--r--  1 apache apache 2.1K Aug 14 05:38 index.php
-rw-r--r--  1 apache apache 2.6K Aug 14 05:38 indie.json
drwxr-xr-x  3 apache apache 4.0K Aug 14 05:38 l10n
drwxr-xr-x  6 apache apache 4.0K Aug 14 05:38 lib
-rw-r--r--  1 apache apache  283 Aug 14 05:38 occ
drwxr-xr-x  2 apache apache 4.0K Aug 14 05:38 ocs
drwxr-xr-x  2 apache apache 4.0K Aug 14 05:38 ocs-provider
-rw-r--r--  1 apache apache 3.0K Aug 14 05:38 public.php
-rw-r--r--  1 apache apache 4.3K Aug 14 05:38 remote.php
-rw-r--r--  1 apache apache   26 Aug 14 05:38 robots.txt
drwxr-xr-x 13 apache apache 4.0K Aug 14 05:38 settings
-rw-r--r--  1 apache apache 1.7K Aug 14 05:38 status.php
drwxr-xr-x  3 apache apache 4.0K Aug 14 05:38 themes
-rw-r--r--  1 apache apache  190 Aug 14 05:38 version.php

The original permissions in owncloud-8.0.4.tar.bz2:
tar -xjvf owncloud-8.0.4.tar.bz2

[user1@oc1 bin]# cd owncloud
[user1@oc1 owncloud]# ls -al
total 164
drwxr-xr-x 11 nobody 65534  4096 Jun  4 08:44 .
drwxr-xr-x  4 user1   user1   4096 Aug 15 09:26 ..
drwxr-xr-x 21 nobody 65534  4096 Jun  4 08:43 3rdparty
drwxrwxrwx 22 nobody 65534  4096 Jun  4 08:44 apps
-rw-r--r--  1 nobody 65534   477 Jun  4 08:43 AUTHORS
drwxrwxrwx  2 nobody 65534  4096 Jun  4 08:43 config
-rw-r--r--  1 nobody 65534  1941 Jun  4 08:43 console.php
-rw-r--r--  1 nobody 65534 34520 Jun  4 08:43 COPYING-AGPL
drwxr-xr-x 19 nobody 65534  4096 Jun  4 08:44 core
-rw-r--r--  1 nobody 65534  5014 Jun  4 08:43 cron.php
-rw-r--r--  1 nobody 65534 23751 Jun  4 08:43 db_structure.xml
-rw-r--r--  1 nobody 65534  1694 Jun  4 08:43 .htaccess
-rw-r--r--  1 nobody 65534   179 Jun  4 08:43 index.html
-rw-r--r--  1 nobody 65534  1792 Jun  4 08:43 index.php
-rw-r--r--  1 nobody 65534  2595 Jun  4 08:43 indie.json
drwxr-xr-x  2 nobody 65534  4096 Jun  4 08:43 l10n
drwxr-xr-x  6 nobody 65534  4096 Jun  4 08:43 lib
-rw-r--r--  1 nobody 65534   283 Jun  4 08:43 occ
drwxr-xr-x  2 nobody 65534  4096 Jun  4 08:43 ocs
-rw-r--r--  1 nobody 65534  1838 Jun  4 08:43 public.php
-rw-r--r--  1 nobody 65534  1820 Jun  4 08:43 remote.php
-rw-r--r--  1 nobody 65534    26 Jun  4 08:43 robots.txt
drwxr-xr-x 13 nobody 65534  4096 Jun  4 08:43 settings
-rw-r--r--  1 nobody 65534  1525 Jun  4 08:43 status.php
drwxr-xr-x  2 nobody 65534  4096 Jun  4 08:43 themes
-rw-r--r--  1 nobody 65534   165 Jun  4 08:43 .user.ini
-rw-r--r--  1 nobody 65534   172 Jun  4 08:46 version.php

HTH.

**Comment by MarkWiseReynoldsburg** [comment]
I found that commenting out the php.ini open_basedir option cleared this up for me.

**Comment by PVince81** [comment]
Does open_basedir fix the problem for the other people here @ymybe @Pingumann @cmeans ?

If yes then maybe the permissions check for the temp dir could log an error and hint at "open_basedir" to tell admins they should disable it ? @LukasReschke

**Comment by cmeans** [comment]
I did get a solution here that allowed me to enable apps, but I don't
recall what it was.  Unfortunately, the data drive was later corrupted and
there was no simple solution for fixing the install (and just
reinitializing the data drive) without a complete reinstall so I've dropped
owncloud for now.  May come back to it when I have more time.

**Comment by cmonteroluque** [comment]
This issue will fall off 8.2 if not resolved by 6 Oct

**Comment by Mickael-Martin** [comment]
Resolved by 8.2 on my RPI2

**Comment by PVince81** [comment]
Please try again with 8.2.5 or later. If this issue still occurs, please reopen
